src/core/qgsvectorfilewriter.cpp

```diff
@@ -650,7 +650,16 @@ void QgsVectorFileWriter::init( QString vectorFileName,
             break;
 
           case QVariant::Double:
-            ogrType = OFTReal;
+            if ( ( mOgrDriverName == QLatin1String( "GPKG" ) || mOgrDriverName == QLatin1String( "SQLite" ) ) &&
+                 attrField.precision() == 0 && attrField.name().compare( QStringLiteral( "fid" ), Qt::CaseInsensitive ) == 0 )
```

Comment:
Instead of comparing to "fid", we should compare to the value of the FID layer creation option (which defaults to "fid")

Reply:
@rouault , fair point.